node/core/av-store/src/lib.rs

```diff
@@ -19,52 +19,290 @@
+///  b) There are records but all of them are in `Included` state and do not have exact time to
+///     be pruned.
+#[derive(Decode, Encode)]
+struct NextChunkPruning(u64);
+
+impl NextChunkPruning {
+	// After which amount of seconds into the future from `now` this should fire.
+	fn should_fire_in(&self) -> Result<u64, Error> {
+		Ok(self.0.saturating_sub(SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs()))
+	}
+}
+
+/// Struct holding pruning timing configuration.
+/// The only purpose of this structure is to use different timing
+/// configurations in production and in testing.
+#[derive(Clone)]
+struct PruningConfig {
+	/// How long should a stored block stay available.
+	keep_stored_block_for: u64,
+
+	/// How long should a finalized block stay available.
+	keep_finalized_block_for: u64,
+
+	/// How long should a chunk of a finalized block stay available.
+	keep_finalized_chunk_for: u64
+}
+
+impl Default for PruningConfig {
+    fn default() -> Self {
+		Self {
+			keep_stored_block_for: KEEP_STORED_BLOCK_FOR,
+			keep_finalized_block_for: KEEP_FINALIZED_BLOCK_FOR,
+			keep_finalized_chunk_for: KEEP_FINALIZED_CHUNK_FOR,
+		}
+    }
+}
+
+#[derive(Debug, Decode, Encode, Eq, PartialEq)]
+enum CandidateState {
+	Stored,
+	Included,
+	Finalized,
+}
+
+#[derive(Debug, Decode, Encode, Eq)]
+struct PoVPruningRecord {
+	candidate_hash: Hash,
+	block_number: BlockNumber,
+	candidate_state: CandidateState,
+	prune_at: u64,
+}
+
+impl PartialEq for PoVPruningRecord {
+	fn eq(&self, other: &Self) -> bool {
+		self.candidate_hash == other.candidate_hash
+	}
+}
+
+impl Ord for PoVPruningRecord {
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.prune_at.cmp(&other.prune_at)
+	}
+}
+
+impl PartialOrd for PoVPruningRecord {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+    }
+}
+
+#[derive(Debug, Decode, Encode, Eq)]
+struct ChunkPruningRecord {
+	candidate_hash: Hash,
+	block_number: BlockNumber,
+	candidate_state: CandidateState,
+	chunk_index: u32,
+	prune_at: u64,
+}
+
+impl PartialEq for ChunkPruningRecord {
+	fn eq(&self, other: &Self) -> bool {
+		self.candidate_hash == other.candidate_hash &&
+			self.chunk_index == other.chunk_index
+	}
+}
+
+impl Ord for ChunkPruningRecord {
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.prune_at.cmp(&other.prune_at)
+	}
+}
+
+impl PartialOrd for ChunkPruningRecord {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
 }
 
 /// An implementation of the Availability Store subsystem.
 pub struct AvailabilityStoreSubsystem {
+	pruning_config: PruningConfig,
 	inner: Arc<dyn KeyValueDB>,
 	metrics: Metrics,
 }
 
+impl AvailabilityStoreSubsystem {
+	// Perform pruning of PoVs
+	fn prune_povs(&self) -> Result<(), Error> {
+		let mut tx = DBTransaction::new();
+		let mut pov_pruning = pov_pruning(&self.inner).unwrap_or_default();
+		let now = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs();
```

Comment:
It sounds like this would not be the only place to suffer in this case; each block has a timestamp extrinsic for instance so you sort of are not supposed to have wrong time?

Reply:
Fair enough. But even the "right" `SystemTime` is allowed to go backwards sometimes and I wonder if it can cause any issues here.

Reply:
If the offset is a matter of seconds, then I can't really see any real problems with pruning a few seconds early / late. 

Changing time zone shouldn't ever be a matter of hours because IIRC we're running NTP on nodes, and handle everything internally in UTC. Still, even if that somehow happens, I think (low confidence, though) that there is no way for a node to lose stake by pruning early/late.

Reply:
FWIW, a timezone has no influence on the UNIX time (leap seconds do though). That doesn't matter since, as @ordian already mentioned, e.g. time sync can lead to skewing time. In general, AFAIK, we cannot rely on the wall clock.

While I share @ordian 's concern, I am not sure if we can use a monotonic clock. Maximum that we can get here, AFAIU, is uptime and I suspect that doesn't work for us due to possible reboots.

So I guess we have to deal with `SystemTime` and make sure that the code we write here is tolerable to small jumps back and forth.

Reply:
Yeah, I guess if the small jumps are considerably smaller than 24 hours, it's fine.
https://w3f.github.io/parachain-implementers-guide/node/utility/availability-store.html

Reply:
>  I suspect that doesn't work for us due to possible reboots.
Yes, that is absolutely true